File: elixir/functions/pattern_matching.ex
defmodule PatternMatching do
  @moduledoc """
  Pattern matching and function examples in Elixir
  """

  # Pattern matching in function heads
  def factorial(0), do: 1
  def factorial(n) when n > 0, do: n * factorial(n - 1)

  # Pattern matching with lists
  def sum_list([]), do: 0
  def sum_list([head | tail]), do: head + sum_list(tail)

  # Pattern matching with tuples
  def describe_point({0, 0}), do: "Origin"
  def describe_point({0, y}), do: "On Y-axis at #{y}"
  def describe_point({x, 0}), do: "On X-axis at #{x}"
  def describe_point({x, y}), do: "Point at (#{x}, #{y})"

  # Pattern matching with maps
  def greet(%{name: name, age: age}) do
    "Hello #{name}, you are #{age} years old"
  end

  def greet(%{name: name}) do
    "Hello #{name}"
  end

  # Case expressions
  def check_value(value) do
    case value do
      0 -> "zero"
      n when n < 0 -> "negative"
      n when n > 0 -> "positive"
      _ -> "unknown"
    end
  end

  # Cond expressions (like else-if chains)
  def grade_score(score) do
    cond do
      score >= 90 -> "A"
      score >= 80 -> "B"
      score >= 70 -> "C"
      score >= 60 -> "D"
      true -> "F"
    end
  end

  # Anonymous functions
  def demo_anonymous do
    # Simple anonymous function
    add = fn a, b -> a + b end
    IO.puts("5 + 3 = #{add.(5, 3)}")

    # Short syntax for anonymous functions
    multiply = &(&1 * &2)
    IO.puts("4 * 6 = #{multiply.(4, 6)}")

    # Passing functions as arguments
    numbers = [1, 2, 3, 4, 5]
    doubled = Enum.map(numbers, fn x -> x * 2 end)
    IO.puts("Doubled: #{inspect(doubled)}")
  end

  # Higher-order functions
  def apply_twice(f, x) do
    f.(f.(x))
  end

  # Pipe operator
  def demo_pipe do
    result =
      "hello world"
      |> String.upcase()
      |> String.split()
      |> Enum.reverse()
      |> Enum.join(" ")

    IO.puts("Piped result: #{result}")
  end

  # Demo function
  def demo do
    IO.puts("=== Factorial ===")
    IO.puts("5! = #{factorial(5)}")

    IO.puts("\n=== Sum List ===")
    IO.puts("Sum of [1,2,3,4,5] = #{sum_list([1, 2, 3, 4, 5])}")

    IO.puts("\n=== Describe Point ===")
    IO.puts(describe_point({0, 0}))
    IO.puts(describe_point({0, 5}))
    IO.puts(describe_point({3, 0}))
    IO.puts(describe_point({3, 4}))

    IO.puts("\n=== Greet ===")
    IO.puts(greet(%{name: "Alice", age: 30}))
    IO.puts(greet(%{name: "Bob"}))

    IO.puts("\n=== Check Value ===")
    IO.puts("0 is #{check_value(0)}")
    IO.puts("-5 is #{check_value(-5)}")
    IO.puts("10 is #{check_value(10)}")

    IO.puts("\n=== Grade Score ===")
    IO.puts("Score 95: #{grade_score(95)}")
    IO.puts("Score 72: #{grade_score(72)}")
    IO.puts("Score 55: #{grade_score(55)}")

    IO.puts("\n=== Anonymous Functions ===")
    demo_anonymous()

    IO.puts("\n=== Higher Order Functions ===")
    increment = fn x -> x + 1 end
    IO.puts("Apply increment twice to 5: #{apply_twice(increment, 5)}")

    IO.puts("\n=== Pipe Operator ===")
    demo_pipe()
  end
end

# Run the demo
PatternMatching.demo()

File: fsharp/functions/pattern_matching.fs
// Pattern matching and discriminated unions in F#

module PatternMatching

open System

// Discriminated unions
type Shape =
    | Circle of radius: float
    | Rectangle of width: float * height: float
    | Triangle of base_: float * height: float

type Option<'T> =
    | Some of 'T
    | None

type Result<'T, 'E> =
    | Ok of 'T
    | Error of 'E

// Pattern matching on shapes
let area shape =
    match shape with
    | Circle radius -> Math.PI * radius * radius
    | Rectangle (width, height) -> width * height
    | Triangle (base_, height) -> 0.5 * base_ * height

// Pattern matching with guards
let describe number =
    match number with
    | n when n < 0 -> "negative"
    | 0 -> "zero"
    | n when n > 0 -> "positive"
    | _ -> "unknown"

// Record types
type Person = {
    Name: string
    Age: int
    City: string
}

// Pattern matching on records
let greet person =
    match person with
    | { Name = name; Age = age } when age < 18 ->
        sprintf "%s is a minor" name
    | { Name = name; Age = age } ->
        sprintf "%s is %d years old" name age

// Active patterns
let (|Even|Odd|) n =
    if n % 2 = 0 then Even else Odd

let (|Positive|Negative|Zero|) n =
    if n > 0 then Positive
    elif n < 0 then Negative
    else Zero

// Recursive pattern matching
let rec factorial n =
    match n with
    | 0 | 1 -> 1
    | n -> n * factorial (n - 1)

let rec fibonacci n =
    match n with
    | 0 -> 0
    | 1 -> 1
    | n -> fibonacci (n - 1) + fibonacci (n - 2)

// Demo
let demo () =
    printfn "\n=== Pattern Matching in F# ===\n"

    // Shape pattern matching
    printfn "=== Shapes ==="
    let circle = Circle 5.0
    let rectangle = Rectangle (4.0, 6.0)
    let triangle = Triangle (3.0, 4.0)

    printfn "Circle area: %.2f" (area circle)
    printfn "Rectangle area: %.2f" (area rectangle)
    printfn "Triangle area: %.2f" (area triangle)

    // Number description
    printfn "\n=== Number Description ==="
    printfn "Describe 0: %s" (describe 0)
    printfn "Describe -5: %s" (describe -5)
    printfn "Describe 10: %s" (describe 10)

    // Record pattern matching
    printfn "\n=== Records ==="
    let person1 = { Name = "Alice"; Age = 30; City = "NYC" }
    let person2 = { Name = "Bob"; Age = 15; City = "LA" }
    printfn "%s" (greet person1)
    printfn "%s" (greet person2)

    // Active patterns
    printfn "\n=== Active Patterns ==="
    let checkParity n =
        match n with
        | Even -> sprintf "%d is even" n
        | Odd -> sprintf "%d is odd" n

    printfn "%s" (checkParity 4)
    printfn "%s" (checkParity 7)

    let checkSign n =
        match n with
        | Positive -> sprintf "%d is positive" n
        | Negative -> sprintf "%d is negative" n
        | Zero -> "zero"

    printfn "%s" (checkSign 5)
    printfn "%s" (checkSign -3)
    printfn "%s" (checkSign 0)

    // Recursion
    printfn "\n=== Recursion ==="
    printfn "Factorial 5: %d" (factorial 5)
    printfn "Fibonacci 10: %d" (fibonacci 10)

    // List pattern matching
    printfn "\n=== List Patterns ==="
    let rec sumList lst =
        match lst with
        | [] -> 0
        | head :: tail -> head + sumList tail

    printfn "Sum [1..5]: %d" (sumList [1..5])

[<EntryPoint>]
let main argv =
    demo()
    0
